### PR TITLE
SchemaBasedMessagesSerdeProvider abstracts from ParsedSchema

### PR DESCRIPTION
**Pull request by gadomsky**

### CODE REVIEW

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaWithMetadata.scala [hunk 1]
```diff
@@ -14,7 +15,8 @@ case class SchemaWithMetadata(schema: ParsedSchema, id: Int)
 object SchemaWithMetadata{
   def apply(schemaMetadata: SchemaMetadata): SchemaWithMetadata = {
     schemaMetadata.getSchemaType match {
-      case "AVRO" => SchemaWithMetadata(new AvroSchema(schemaMetadata.getSchema), schemaMetadata.getId)
+      case AvroSchema.TYPE => SchemaWithMetadata(new AvroSchema(schemaMetadata.getSchema), schemaMetadata.getId)
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaWithMetadata.scala:18]: I wonder if we can do it somehow "automatically"... e.g. using SchemaRegistryClient.parseSchema?

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/source/UniversalKafkaSourceFactory.scala [hunk 2]
```diff
@@ -27,7 +28,7 @@ import scala.reflect.ClassTag
   * This is universal kafka source - it will handle both avro and json
   * TODO: Move it to some other module when json schema handling will be available
   */
-class UniversalKafkaSourceFactory[K: ClassTag, V: ClassTag](val schemaRegistryProvider: SchemaRegistryProvider,
+class UniversalKafkaSourceFactory[K: ClassTag, V: ClassTag](val schemaRegistryProvider: SchemaRegistryProvider[AvroSchema],
```
Comment (mproch) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/source/UniversalKafkaSourceFactory.scala:31]: this is temporary, right?

**Review by mproch — COMMENTED** [review]
Looks good, let's start to write migration guide for this set of changes ;)

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryProvider.scala [hunk 3]
```diff
@@ -5,13 +5,14 @@ import io.confluent.kafka.schemaregistry.ParsedSchema
 import pl.touk.nussknacker.engine.avro.serialization.{KafkaAvroDeserializationSchemaFactory, KafkaAvroSerializationSchemaFactory}
 import pl.touk.nussknacker.engine.kafka.RecordFormatterFactory
 
-trait SchemaRegistryProvider extends Serializable with BaseSchemaRegistryProvider {
+//todo: rename and separate from BaseSchemaRegistryProvider
+trait SchemaRegistryProvider[T<:ParsedSchema] extends Serializable with BaseSchemaRegistryProvider {
```
Comment (arkadius) [inline utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaRegistryProvider.scala:9]: IMO The genesis of `SchemaRegistryProvider` name was "provider of implementation of mechanisms serializing/deserializing avro records".  You can imagine situations like:
- you have own, file based/configuration based schema registries (two our clients wanted such feature)
- you have other method of avro serailization - e.g. in confluent it is magic byte + 4bytes for schema id + payload, hermes uses schemaid in header, payload in body, I don't know how other schema registries works (e.g. hortonworks schema registry)

We can change the name to more suitable. Maybe: `SchemaBasedMessagesSerdeProvider`?
Good point that it should be separated from `BaseSchemaRegistryProvider` after adding this generic parameter.

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaWithMetadata.scala (hunk 1 above)
Comment (arkadius) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/SchemaWithMetadata.scala:18]: Yeah, we had a conversation about it with @gadomsky - Confluent do exactly the same as we do - they even have their own `nonRestrictiveParseSchema` `(validateDefaults = false)`

File: utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/source/UniversalKafkaSourceFactory.scala (hunk 2 above)
Comment (arkadius) [reply utils/avro-components-utils/src/main/scala/pl/touk/nussknacker/engine/avro/source/UniversalKafkaSourceFactory.scala:31]: IMO we shouldn't have `SchemaRegistryProvider[ParsedSchema]`, we should rather pass here sth like `fun[T<: ParsedSchema](schema: T): SchemaRegistryProvider[T]`. Maybe this PR can show a little bit further direction? Until now it isn't too much code.

**Pull request closed without merging**